### DISCUSSION TITLE
How to know which target is selected for a AOT multitarget static library?

### DISCUSSION THREAD

**Discussion by mcourteaux** [halide/Halide#4924] [General]
How do I know which target it selected, without changing all targets to `*-debug` ?
Also, is there a way to influence the selection at runtime? For example to balance load between CPU and GPU implementations?

I'm planning on generating and gathering performance reports from users, and knowing which version of the pipeline they run would be very interesting. It'd be great if I could simply know, without having to rewrite the dispatch layer myself.

**Comment by steven-johnson** [comment]
> How do I know which target it selected, without changing all targets to `*-debug` ?

There isn't currently an official way to tell which target it selected. But you could hack around it in several ways; most notably, something like:

```
Expr e;
if (get_target().has_feature(Target::AVX2))
    e = 2;
else if (get_target().has_feature(Target::AVX))
    e = 1;
else
    e = 0;
```

If you needed more sophisticated runtime sniffing, you could insert a `define_extern` to call out and do the sniffing for you.

> Also, is there a way to influence the selection at runtime? For example to balance load between CPU and GPU implementations?

Yes and no: You can replace the implementation of `halide_default_can_use_target_features` (either by weak-link or by `halide_set_custom_can_use_target_features`) to make it do whatever you want... but the underlying implementation of multitarget will only call this once (the first time your function is called) and 'lock in' the call for the rest of the process lifetime, for efficiency.

If you need a more robust way to choose different implementations at runtime, I'd strongly recommend just writing your own wrapper code to do so and not trying to use the multitarget facility at all; the multitarget stuff is really targeted at things like the x86 architecture where there is a clear and obvious best choice based on CPUID.
